gobblin-runtime/src/main/java/gobblin/runtime/JobContext.java

```diff
@@ -345,26 +355,48 @@ void finalizeJobStateBeforeCommit() {
    */
   void commit() throws IOException {
     this.datasetStatesByUrns = Optional.of(this.jobState.createDatasetStatesByUrns());
-    boolean allDatasetsCommit = true;
-    boolean shouldCommitDataInJob = shouldCommitDataInJob(this.jobState);
-    DeliverySemantics deliverySemantics = DeliverySemantics.parse(this.jobState);
+    final boolean shouldCommitDataInJob = shouldCommitDataInJob(this.jobState);
+    final DeliverySemantics deliverySemantics = DeliverySemantics.parse(this.jobState);
+    int numCommitThreads = numCommitThreads(this.jobState);
 
     if (!shouldCommitDataInJob) {
       this.logger.info("Job will not commit data since data are committed by tasks.");
     }
 
-    for (Map.Entry<String, JobState.DatasetState> entry : this.datasetStatesByUrns.get().entrySet()) {
-      allDatasetsCommit &=
-          processDatasetCommit(shouldCommitDataInJob, deliverySemantics, entry.getKey(), entry.getValue());
+    try {
+      List<Either<Void, ExecutionException>> result =
+          new IteratorExecutor<>(Iterables.transform(this.datasetStatesByUrns.get().entrySet(),
+          new Function<Map.Entry<String, DatasetState>, Callable<Void>>() {
+            @Nullable
+            @Override
+            public Callable<Void> apply(@Nullable final Map.Entry<String, DatasetState> entry) {
+              return new Callable<Void>() {
+                @Override
+                public Void call()
+                    throws Exception {
+                  processDatasetCommit(shouldCommitDataInJob, deliverySemantics, entry.getKey(), entry.getValue());
```

Comment:
Please also add some unit tests, e.g. to verify if failures are handled correctly.